File: src/responder/registration.cpp
#include "opaque/responder.h"
#include "opaque/protocol.h"
#include "opaque/pq.h"
#include "opaque/debug_log.h"
#include <sodium.h>
#include <algorithm>

namespace ecliptix::security::opaque::responder {
    namespace {
        namespace oblivious_prf = oblivious_prf;
        namespace crypto = crypto;
    }

    RegistrationResponse::RegistrationResponse() : data(REGISTRATION_RESPONSE_LENGTH) {
    }

    Result create_registration_response_impl(const uint8_t *registration_request, const size_t request_length,
                                             const secure_bytes &responder_private_key,
                                             const secure_bytes &responder_public_key,
                                             const uint8_t *account_id,
                                             const size_t account_id_length,
                                             RegistrationResponse &response) {
        log::section("RELAY: Create Registration Response");
        if (!registration_request || request_length != REGISTRATION_REQUEST_LENGTH ||
            !account_id || account_id_length == 0) [[unlikely]] {
            return Result::InvalidInput;
        }
        if (!crypto::init()) {
            return Result::CryptoError;
        }
        log::hex("registration_request (blinded element)", registration_request, request_length);
        log::hex("responder_private_key", responder_private_key);
        log::hex("responder_public_key", responder_public_key);
        log::hex("account_id", account_id, account_id_length);

        const uint8_t *blinded_element = registration_request;
        if (Result point_result = crypto::validate_ristretto_point(blinded_element, REGISTRATION_REQUEST_LENGTH);
            point_result != Result::Success) {
            return Result::InvalidInput;
        }
        uint8_t evaluated_element[crypto_core_ristretto255_BYTES];
        uint8_t oprf_key[PRIVATE_KEY_LENGTH] = {};
        Result result = crypto::derive_oprf_key(responder_private_key.data(), responder_private_key.size(),
                                                account_id, account_id_length, oprf_key);
        if (result != Result::Success) [[unlikely]] {
            sodium_memzero(oprf_key, sizeof(oprf_key));
            return result;
        }
        log::hex("oprf_key (derived from responder_private_key + account_id)", oprf_key, sizeof(oprf_key));
        result = oblivious_prf::evaluate(blinded_element, oprf_key, evaluated_element);
        sodium_memzero(oprf_key, sizeof(oprf_key));
        if (result != Result::Success) [[unlikely]] {
            return result;
        }
        log::hex("evaluated_element (OPRF output)", evaluated_element, sizeof(evaluated_element));

        size_t offset = 0;
        std::copy_n(evaluated_element, crypto_core_ristretto255_BYTES,
                    response.data.begin() + static_cast<std::ptrdiff_t>(offset));
        offset += crypto_core_ristretto255_BYTES;
        std::ranges::copy(responder_public_key,
                          response.data.begin() + static_cast<std::ptrdiff_t>(offset));
        log::hex("registration_response", response.data);
        return Result::Success;
    }

    Result build_credentials(const uint8_t *registration_record, size_t record_length,
                             ResponderCredentials &credentials) {
        log::section("RELAY: Build Credentials from Registration Record");
        const size_t record_expected = REGISTRATION_RECORD_LENGTH;
        if (!registration_record || record_length < record_expected) {
            return Result::InvalidInput;
        }
        if (!crypto::init()) {
            return Result::CryptoError;
        }
        log::hex("registration_record", registration_record, record_length);
        protocol::RegistrationRecordView record_view{};
        Result parse_result = protocol::parse_registration_record(registration_record, record_length, record_view);
        if (parse_result != Result::Success) {
            return parse_result;
        }
        if (Result key_result = crypto::validate_public_key(record_view.initiator_public_key, PUBLIC_KEY_LENGTH);
            key_result != Result::Success) {
            return key_result;
        }

        credentials.envelope.assign(record_view.envelope, record_view.envelope + ENVELOPE_LENGTH);
        credentials.initiator_public_key.assign(record_view.initiator_public_key,
                                                record_view.initiator_public_key + PUBLIC_KEY_LENGTH);
        log::hex("credentials.envelope", credentials.envelope);
        log::hex("credentials.initiator_public_key (EC)", credentials.initiator_public_key);
        return Result::Success;
    }
}

File: rust/crates/opaque-relay/src/registration.rs
// Copyright (c) 2026 Oleksandr Melnychenko, Ukraine
// Ecliptix Security — Hybrid PQ-OPAQUE
// Licensed under the MIT License

use opaque_core::types::{
    OpaqueError, OpaqueResult, PRIVATE_KEY_LENGTH, PUBLIC_KEY_LENGTH, REGISTRATION_REQUEST_LENGTH,
};
use opaque_core::{crypto, oprf, protocol};
use zeroize::Zeroize;

use crate::state::{OpaqueResponder, RegistrationResponse, ResponderCredentials};

/// Evaluates the OPRF on the blinded request and builds a registration response.
///
/// Derives a per-account OPRF key from the relay OPRF seed and `account_id`,
/// evaluates the blinded element from `registration_request`, and writes the
/// result together with the relay public key into `response`.
///
/// # Errors
///
/// Returns an error if `registration_request` has an invalid length, is not a
/// valid Ristretto255 point, `account_id` is empty, or the OPRF evaluation fails.
pub fn create_registration_response(
    responder: &OpaqueResponder,
    registration_request: &[u8],
    account_id: &[u8],
    response: &mut RegistrationResponse,
) -> OpaqueResult<()> {
    if registration_request.len() != REGISTRATION_REQUEST_LENGTH {
        return Err(OpaqueError::InvalidProtocolMessage);
    }
    if account_id.is_empty() {
        return Err(OpaqueError::InvalidInput);
    }

    crypto::validate_ristretto_point(registration_request)?;

    let mut oprf_key = [0u8; PRIVATE_KEY_LENGTH];
    crypto::derive_oprf_key(
        responder.oprf_seed(),
        account_id,
        &mut oprf_key,
    )?;

    let blinded: &[u8; PUBLIC_KEY_LENGTH] = registration_request
        .try_into()
        .map_err(|_| OpaqueError::InvalidProtocolMessage)?;
    let mut evaluated = [0u8; PUBLIC_KEY_LENGTH];
    oprf::evaluate(blinded, &oprf_key, &mut evaluated)?;
    oprf_key.zeroize();

    response.data[..PUBLIC_KEY_LENGTH].copy_from_slice(&evaluated);
    response.data[PUBLIC_KEY_LENGTH..].copy_from_slice(responder.public_key());

    Ok(())
}

/// Parses a registration record and stores it in the relay credential file.
///
/// Extracts the envelope and the initiator public key from `registration_record`
/// and writes them into `credentials`. Refuses to overwrite previously stored
/// credentials.
///
/// # Errors
///
/// Returns [`OpaqueError::AlreadyRegistered`] if credentials have already been
/// populated, or an error if the record is malformed or the initiator public key
/// is invalid.
pub fn build_credentials(
    registration_record: &[u8],
    credentials: &mut ResponderCredentials,
) -> OpaqueResult<()> {
    if credentials.registered {
        return Err(OpaqueError::AlreadyRegistered);
    }

    let view = protocol::parse_registration_record(registration_record)?;

    crypto::validate_public_key(view.initiator_public_key)?;

    credentials.envelope = view.envelope.to_vec();
    credentials
        .initiator_public_key
        .copy_from_slice(view.initiator_public_key);
    credentials.registered = true;

    Ok(())
}
